File: src/jira_api/services/project_service.py
"""Project service for JIRA operations."""

from typing import List, Optional

from jira_api.core.client import JiraClient
from jira_api.models.project import Project, ProjectVersion, ProjectVersionCreate


class ProjectService:
    """Service for project-related operations."""

    def __init__(self, client: JiraClient) -> None:
        """Initialize the project service.

        Args:
            client: JIRA API client instance
        """
        self.client = client

    def get_project(self, project_key: str) -> Project:
        """Get a project by its key.

        Args:
            project_key: The project key

        Returns:
            Project object
        """
        return self.client.get_project(project_key)

    def get_project_versions(
        self, 
        project_key: str, 
        released_only: Optional[bool] = None
    ) -> List[ProjectVersion]:
        """Get all versions for a project.

        Args:
            project_key: The project key
            released_only: If True, return only released versions.
                          If False, return only unreleased versions.
                          If None, return all versions.

        Returns:
            List of ProjectVersion objects
        """
        versions = self.client.get_project_versions(project_key)
        
        if released_only is None:
            return versions
        elif released_only:
            return [v for v in versions if v.released]
        else:
            return [v for v in versions if not v.released]

    def create_version(
        self,
        project_key: str,
        version_name: str,
        description: Optional[str] = None,
        start_date: Optional[str] = None,
        release_date: Optional[str] = None,
        released: bool = False,
        archived: bool = False,
    ) -> ProjectVersion:
        """Create a new version for a project.

        Args:
            project_key: The project key
            version_name: Name of the version
            description: Optional description
            start_date: Optional start date (YYYY-MM-DD format)
            release_date: Optional release date (YYYY-MM-DD format)
            released: Whether the version is released
            archived: Whether the version is archived

        Returns:
            Created ProjectVersion object
        """
        # First get the project to obtain its ID
        project = self.get_project(project_key)
        project_id = int(project.id)

        version_data = ProjectVersionCreate(
            name=version_name,
            description=description,
            project_id=project_id,
            start_date=start_date,
            release_date=release_date,
            released=released,
            archived=archived,
        )

        return self.client.create_project_version(version_data)

    def get_version_by_name(self, project_key: str, version_name: str) -> Optional[ProjectVersion]:
        """Get a specific version by name.

        Args:
            project_key: The project key
            version_name: Name of the version to find

        Returns:
            ProjectVersion object if found, None otherwise
        """
        versions = self.get_project_versions(project_key)
        
        for version in versions:
            if version.name == version_name:
                return version
        
        return None

    def get_released_versions(self, project_key: str) -> List[ProjectVersion]:
        """Get all released versions for a project.

        Args:
            project_key: The project key

        Returns:
            List of released ProjectVersion objects
        """
        return self.get_project_versions(project_key, released_only=True)

    def get_unreleased_versions(self, project_key: str) -> List[ProjectVersion]:
        """Get all unreleased versions for a project.

        Args:
            project_key: The project key

        Returns:
            List of unreleased ProjectVersion objects
        """
        return self.get_project_versions(project_key, released_only=False)

    def get_issue_types(self) -> List[dict]:
        """Get all available issue types.

        Returns:
            List of issue type data
        """
        return self.client.get_issue_types()

File: node/src/services/project.service.ts
/**
 * ProjectService - High-level project operations
 * Provides project metadata, version management, and issue type helpers
 */

import type { JiraClient } from '../core/client.js';
import type {
  Project,
  ProjectVersion,
  IssueType,
  VersionQuery,
  VersionCreateInput,
} from '../models/project.js';
import { JiraNotFoundError, JiraValidationError } from '../errors/index.js';
import { getLogger } from '../utils/logger.js';

/**
 * ProjectService class
 * Provides business logic layer for project operations
 */
export class ProjectService {
  private logger = getLogger();

  constructor(private client: JiraClient) {}

  // ======================
  // PROJECT METHODS
  // ======================

  /**
   * Get project by key
   */
  async getProject(key: string): Promise<Project> {
    this.logger.debug({ key }, 'Getting project');
    return this.client.getProject(key);
  }

  // ======================
  // VERSION METHODS
  // ======================

  /**
   * Get project versions
   */
  async getProjectVersions(
    projectKey: string,
    filter?: 'released' | 'unreleased'
  ): Promise<ProjectVersion[]> {
    this.logger.debug({ projectKey, filter }, 'Getting project versions');

    const query: VersionQuery = {};
    if (filter === 'released') {
      query.released = true;
    } else if (filter === 'unreleased') {
      query.released = false;
    }

    return this.client.getProjectVersions(projectKey, query);
  }

  /**
   * Create version
   */
  async createVersion(
    projectKey: string,
    data: VersionCreateInput
  ): Promise<ProjectVersion> {
    this.logger.debug({ projectKey, data }, 'Creating version');

    // Validate required fields
    if (!data.name) {
      throw new JiraValidationError('Version name is required');
    }

    // Get project to get project ID
    const project = await this.client.getProject(projectKey);

    return this.client.createProjectVersion(projectKey, {
      ...data,
      projectId: parseInt(project.id, 10),
    });
  }

  /**
   * Get version by name
   */
  async getVersionByName(
    projectKey: string,
    versionName: string
  ): Promise<ProjectVersion> {
    this.logger.debug({ projectKey, versionName }, 'Getting version by name');

    const versions = await this.getProjectVersions(projectKey);
    const version = versions.find(
      (v) => v.name.toLowerCase() === versionName.toLowerCase()
    );

    if (!version) {
      const availableVersions = versions.map((v) => v.name).join(', ');
      throw new JiraNotFoundError(
        `Version '${versionName}' not found in project ${projectKey}. ` +
          `Available: ${availableVersions || 'none'}`
      );
    }

    return version;
  }

  /**
   * Get released versions
   */
  async getReleasedVersions(projectKey: string): Promise<ProjectVersion[]> {
    return this.getProjectVersions(projectKey, 'released');
  }

  /**
   * Get unreleased versions
   */
  async getUnreleasedVersions(projectKey: string): Promise<ProjectVersion[]> {
    return this.getProjectVersions(projectKey, 'unreleased');
  }

  // ======================
  // ISSUE TYPE METHODS
  // ======================

  /**
   * Get issue types for project
   */
  async getIssueTypes(projectKey: string): Promise<IssueType[]> {
    this.logger.debug({ projectKey }, 'Getting issue types');
    return this.client.getProjectIssueTypes(projectKey);
  }

  /**
   * Get issue type by name
   */
  async getIssueTypeByName(projectKey: string, typeName: string): Promise<IssueType> {
    this.logger.debug({ projectKey, typeName }, 'Getting issue type by name');

    const issueTypes = await this.getIssueTypes(projectKey);
    const issueType = issueTypes.find(
      (type) => type.name.toLowerCase() === typeName.toLowerCase()
    );

    if (!issueType) {
      const availableTypes = issueTypes.map((t) => t.name).join(', ');
      throw new JiraNotFoundError(
        `Issue type '${typeName}' not found in project ${projectKey}. ` +
          `Available: ${availableTypes}`
      );
    }

    return issueType;
  }

  /**
   * Get issue type ID by name
   */
  async getIssueTypeIdByName(projectKey: string, typeName: string): Promise<string> {
    const issueType = await this.getIssueTypeByName(projectKey, typeName);
    return issueType.id;
  }

  /**
   * Check if issue type exists in project
   */
  async issueTypeExists(projectKey: string, typeName: string): Promise<boolean> {
    try {
      await this.getIssueTypeByName(projectKey, typeName);
      return true;
    } catch (error) {
      if (error instanceof JiraNotFoundError) {
        return false;
      }
      throw error;
    }
  }

  /**
   * Get all available issue type names for a project
   */
  async getIssueTypeNames(projectKey: string): Promise<string[]> {
    const issueTypes = await this.getIssueTypes(projectKey);
    return issueTypes.map((type) => type.name);
  }
}
